File: 4- Find the duplicate/Java/solution3.java
// By using a dictionary:

// Time complexity: O(n)
// Space complexity: O(n)

static int findDuplicate(int[] arr){
  HashMap<Integer,Boolean> visited = new HashMap();
  for(int element : arr){
    if(visited.containsKey(element)) return element;
    else visited.put(element, true);
  }
  return 0;
}

File: 4- Find the duplicate/C++/solution3.cpp
// By using a dictionary:

// Time complexity: O(n)
// Space complexity: O(n)

int findDuplicate(vector<int> arr){
  unordered_map<int,bool> visited;
  for(int i = 0; i < arr.size(); i++){
    if(visited.find(arr[i]) != visited.end()) return arr[i];
    else visited[arr[i]] = true;
  }
  return 0;
}


